### PR TITLE
Cython lint

### PR DESCRIPTION
**Pull request by stefanv**

### CODE REVIEW

**Comment by j-bowhay** [comment]
Thanks, looks like this addresses the issues I mentioned :)

**Comment by stefanv** [comment]
The failure here is the same one as on master; looks like on macOS we are installing via conda but without the conda-forge channel?

**Comment by tupui** [comment]
FYI I am looking into this, I will make a PR to attempt a fix.

**Comment by tupui** [comment]
It's fixed, I am rerunning your job (should pickup the update)

File: scipy/io/matlab/_mio5_utils.pyx
Comment (tupui) [inline scipy/io/matlab/_mio5_utils.pyx:158]: Does not look right no?

File: scipy/special/_test_internal.pyx
Comment (tupui) [inline scipy/special/_test_internal.pyx:11]: haha yes although we want to get rid of `assert_`

**Review by tupui — COMMENTED** [review]
Thanks! I just noted something with the classes' doc. Otherwise LGTM

File: scipy/io/matlab/_mio5_utils.pyx
Comment (stefanv) [reply scipy/io/matlab/_mio5_utils.pyx:158]: Could you explain?

**Comment by stefanv** [comment]
@tupui Still erroring.

File: scipy/spatial/_ckdtree.pyx
Comment (WarrenWeckesser) [inline scipy/spatial/_ckdtree.pyx:1549]: Just curious: why does this need `# no-cython-lint`?

Comment (stefanv) [reply scipy/spatial/_ckdtree.pyx:1549]: Good eyes ;)

https://github.com/MarcoGorelli/cython-lint/issues/52

File: scipy/io/matlab/_streams.pyx
Comment (WarrenWeckesser) [inline scipy/io/matlab/_streams.pyx:231]: Is the use of `# no-cython-lint` here also because of https://github.com/MarcoGorelli/cython-lint/issues/52?

Comment (stefanv) [reply scipy/io/matlab/_streams.pyx:231]: It was because of https://github.com/MarcoGorelli/cython-lint/issues/50, but can be removed with the latest cython-lint.

Comment (stefanv) [reply scipy/io/matlab/_streams.pyx:231]: Actually, it's good you are taking a look at this. I wasn't sure whether it is safe, in this specific instance, to not assign to obj (which is what the linter wants).

Comment (WarrenWeckesser) [reply scipy/io/matlab/_streams.pyx:231]: I should have looked at the full code before commenting... in this case, `obj` is, in fact, not explicitly used.  However, the test fails if the result is not assigned to something, probably because of a memory lifetime issue.  Instead of adding `# no-cython-lint`, the widely-used linter convention of prefixing the name with `_` could be used.

I also wonder if there is any point to declaring the object as `cdef object`.  In Cython is there any difference between `cdef object x = ...` and just `x = ...`? (That's not a rhetorical question--I don't know.)  The test passes and the linter does not complain if I use

```
    _obj = st.read_string(n, &d_ptr, True)
```
but if you want these changes to be minimal--just what is needed to make cython-lint happy, then
```
    cdef object _obj = st.read_string(n, &d_ptr, True)
```
also works.

File: scipy/io/matlab/_mio5_utils.pyx
Comment (MarcoGorelli) [reply scipy/io/matlab/_mio5_utils.pyx:158]: is this to get around the "pointless string statement" error? If you make this the docstring, then it won't be thrown, i.e.
```
cdef class VarReader5:
    """ Initialize from file reader object
    <rest of docstring>
    """
    cdef public int is_swapped, little_endian
```

(I may need to introduce configuration at some point..)

File: scipy/spatial/_ckdtree.pyx
Comment (MarcoGorelli) [reply scipy/spatial/_ckdtree.pyx:1549]: fixed in version 0.12.1 - thanks for the report, much appreciated!

File: scipy/io/matlab/_mio5_utils.pyx
Comment (MarcoGorelli) [inline scipy/io/matlab/_mio5_utils.pyx:11]: if you write this as a single block, the linter won't complain, it'll recognise it as the docstring
```

''' Cython mio5 utility routines (-*- python -*- like)

Programmer's notes
------------------
Routines here have been reasonably optimized.
The char matrix reading is not very fast, but it's not usually a
bottleneck. See comments in ``read_char`` for possible ways to go if you
want to optimize.
'''
```

Comment (stefanv) [reply scipy/io/matlab/_mio5_utils.pyx:11]: Right; I think the idea was for the programmer's notes not to appear in the module-level docstring, but in this case it probably doesn't matter, given that it's internal.

**Comment by stefanv** [comment]
Before merging, we need https://github.com/MarcoGorelli/cython-lint/issues/56 resolved.

**Comment by stefanv** [comment]
Apparently, despite https://github.com/MarcoGorelli/cython-lint/issues/56, cython-lint returns error code 0, so we may be good to go after all.

**Comment by MarcoGorelli** [comment]
Yeah I should probably clarify the error message to say something like "can't parse <file>, skipping. Error message: <error from cython's function>"
Ideally of course it wouldn't fail to parse - I'll see what can be done about that

**Review by WarrenWeckesser — APPROVED** [review]
Thanks @stefanv, looks good.

**Comment by scottshambaugh** [comment]
It seems like the `--no-pycodestyle` flag is causing us to miss a lot of things that should be linted, try it out both ways on `scipy/spatial/transform/_rotation.pyx` for example.

**Comment by stefanv** [comment]
> It seems like the `--no-pycodestyle` flag is causing us to miss a lot of things that should be linted, try it out both ways on `scipy/spatial/transform/_rotation.pyx` for example.

They're all stylistic though, so :man_shrugging:. Feel free to remove the flag, but then you have to fix all the errors that crop up all over SciPy.

**Comment by scottshambaugh** [comment]
> They're all stylistic though, so 🤷‍♂️. Feel free to remove the flag, but then you have to fix all the errors that crop up all over SciPy.

Ah okay, I was under the impression that the push to 'lint all the things' was to do a big pass over the whole codebase to get it in a clean state but if that's not the case then I'm happy as-is.

**Comment by stefanv** [comment]
I'm trying to find that sweet spot where we add some checking, but where the number of errors are manageable to fix by hand. It's hard to introduce stricter linting, unless you also use an auto-formatter. Autopep8 and pyupgrade helped some with the regular Python code, but unfortunately we don't have similar things for Cython. Also, most of the errors come from too long lines!

**Review by tupui — APPROVED** [review]
Let's get this in. Thanks @stefanv and thanks all for the reviews.

**Pull request merged**
